### ISSUE TITLE
new problem using models with batch normalization layer 

### ISSUE THREAD

**Issue by mortenwp** [Dobiasd/frugally-deep#209]
The code changes that has happened in February, seems to make it impossible to use models with a  batch normalization layer.
A simple example as the one below will output "invalid beta" and exit
The actual place where this happens is when the size of the moving_mean is being compared
to the depth of the input in the batch normalization layer apply to slices function.

Everything works fine if I use a frugally-deep version from end of january

The input is width=336,height=336 and depth=1, but at time I get to apply to slices in the batch normalization height=1,width=336, depth=336. The axis for batch normalization layer seems to be 3
so it looks like there is some logic that fails when the layer tries to permute the dimension.
After permutation the depth layer that was 1 is now 336, so does this imply that we should not compare the with the depth of the input, but rather the value matching the axis (in this case 3)

-Morten

PS: I can try to see if I can compile a tiny model where this problem occur (the one I have is bit fat)

#include <fdeep/fdeep.hpp>
int main()
{

    std::size_t nx(336);
    std::size_t ny(336);

    std::vector<float> v(nx*ny);
    const auto model = fdeep::load_model("model.json",false);

    const auto shapes= model.get_input_shapes();
    std::cerr << fdeep::show_tensor_shape_variable(shapes[0]) << std::endl;

    std::cin.read(reinterpret_cast<char*>(&v[0]),nx*ny*sizeof(float));

    //for(int i=0;i<100;++i)
     {
    const fdeep::shared_float_vec sv(fplus::make_shared_ref<fdeep::float_vec>(v));
    fdeep::tensor_shape shp(1,1,nx,ny,1);
    fdeep::tensor t(shp, sv);
    //std::cerr << fdeep::show_tensor5(t) << std::endl;
    const auto result = model.predict({t});
    const auto resultd=tensor_to_depth_slices(result[0]);
    for(std::size_t i=0;i<3;++i)
      {
       const std::vector<float> w=*resultd[i].as_vector();
       std::cerr << w.size() << std::endl;
       std::cout.write(reinterpret_cast<const char*>(&(w[0])),nx*ny*sizeof(float));
     }
}

}

**Comment by Dobiasd** [comment]
Hi Morten, thanks for the report.

Since the tests already cover `BatchNormalization` heavily (https://github.com/Dobiasd/frugally-deep/blob/dbaad7a9a3711f5fd7f6c54f4274fbcd4987086f/keras_export/generate_test_models.py#L159-L200), a (minimal) example to reproduce the problem locally would be very helpul.

**Comment by mortenwp** [comment]
Thanks, I just discovered the #206  issue and has been reading through that.
I will see if I get around to do it within the next couple of days.
The above tests models scripts will probably help me :)

As said my input 336,336,1 -> after permutation that becomes 1,336,336  -> and when I compare
moving_mean.size() with  input.shape().depth_ the first one is 1 and the second is 336.
So either moving_mean  should have the size of 336, or we should not compare with the depths.

But yes, minimal example was what you asked for, so I will get back to you :)

And before I forget it, thanks for the rapid response 

-Morten

**Comment by Dobiasd** [comment]
OK, I was able to re-create your issue with the following minimal example:

```python
# fdeep_issue_209.py
import numpy as np
from tensorflow.keras.layers import Input, Permute, BatchNormalization
from tensorflow.keras.models import Model

inputs = Input(shape=(9, 16, 1))
x = Permute((3, 1, 2))(inputs)
x = BatchNormalization()(x)
model = Model(inputs=inputs, outputs=x)

model.compile(loss='categorical_crossentropy', optimizer='nadam')
model.save('fdeep_issue_209.h5', include_optimizer=False)
```

```cpp
// fdeep_issue_209.cpp
#include <fdeep/fdeep.hpp>
int main()
{
    const auto model = fdeep::load_model("fdeep_issue_209.json");
}
```

```bash
g++ -std=c++14 -O3 fdeep_issue_209.cpp -o fdeep_issue_209
python3 fdeep_issue_209.py
python3 convert_model.py fdeep_issue_209.h5 fdeep_issue_209.json
./fdeep_issue_209
```

```
Loading json ... done. elapsed time: 0.000124 s
Building model ... done. elapsed time: 0.000214 s
Running test 1 of 1 ... terminate called after throwing an instance of 'std::runtime_error'
  what():  invalid beta
[1]    9652 abort      ./fdeep_issue_209
```

Seems the permute layer has a bug. I'll fix it and improve test coverage for it.

**Comment by Dobiasd** [comment]
Just pushed a fix. The example from above is working now. Please let me know if this also solved your problem. (You'll need to install the latest version of FunctionalPlus alongside the latest version from `master` of frugally-deep.)

**Comment by mortenwp** [comment]
Thanks for rapid response, but it did not fix the issue.

I think your test tries to test the right thing what happens when you use tensor with depth 1 
as input, and then permute it.

When my batch normalization layer is created the size of moving_mean and friends are all
1 (1 element), but when the appy to slices method is called the input that is being passed to it
are the permuted tensor, that now has a depth of 336 in my case.

The idea with the permuting of the tensor is that we can work on depth dimension always.
If that is the case moving_mean, moving_average (etc) should have a size that matches the depth of the permuted tensor.

If I look in create_batch_normalization_layer, i always get a vector of length of 1. What I dont follow
is that it seems to construct this from base64, but if i look in my model file
I have a moving_mean_initializer that is set to 'Zeros', and some reference to config (which i'm assuming will refer to unpermuted tensor). 

I will try to have a look at the model you generated

best regards
-Morten

-Morten

**Comment by Dobiasd** [comment]
A minimal example to reproduce the problem would be really great. :slightly_smiling_face:

**Comment by mortenwp** [comment]
In your issue209 above test you load a compiled model but  you dont run any predictions.
Also the model run a permutation before running batch normalization, and it does not try to set 
play with the axis parameter that

I have updated the tests and I think the ones below trigger the error (axis=2 works. anything else fails)

`# fdeep_issue_209.py
import numpy as np
from tensorflow.keras.layers import Input, Permute, BatchNormalization
from tensorflow.keras.models import Model

inputs = Input(shape=(9, 16, 1))
#try different axis
x = BatchNormalization(axis=3)(inputs)
model = Model(inputs=inputs, outputs=x)

model.compile(loss='categorical_crossentropy', optimizer='nadam')
model.save('fdeep_issue_209.h5', include_optimizer=False)`

and

cpp code 

// fdeep_issue_209.cpp
#include <fdeep/fdeep.hpp>
int main()
{
    const auto model = fdeep::load_model("fdeep_issue_209.json");
    const auto shapes= model.get_input_shapes();
    std::vector<float> v(9*16);
    const fdeep::shared_float_vec sv(fplus::make_shared_ref<fdeep::float_vec>(v));
    fdeep::tensor_shape shp(9,16,1);
    fdeep::tensor t(shp, sv);
    const auto result = model.predict({t});


}

**Comment by Dobiasd** [comment]
> In your issue209 above test you load a compiled model but you dont run any predictions.

`fdeep::load_model` automatically runs a prediction with the generated test data.

Anyways, the C++ code you posted was not valid (`std::vector v(9*16);`), so here is a corrected version of your minimal example:

```python
# fdeep_issue_209.py
import numpy as np
from tensorflow.keras.layers import Input, Permute, BatchNormalization
from tensorflow.keras.models import Model

inputs = Input(shape=(9, 16, 1))
#try different axis
x = BatchNormalization(axis=3)(inputs)
model = Model(inputs=inputs, outputs=x)

model.compile(loss='categorical_crossentropy', optimizer='nadam')
model.save('fdeep_issue_209.h5', include_optimizer=False)
```

```cpp
// fdeep_issue_209.cpp
#include <fdeep/fdeep.hpp>
int main()
{
    const auto model = fdeep::load_model("fdeep_issue_209.json");
    const auto shapes= model.get_input_shapes();
    std::vector<float> v(9*16);
    const fdeep::shared_float_vec sv(fplus::make_shared_ref<fdeep::float_vec>(v));
    fdeep::tensor_shape shp(9,16,1);
    fdeep::tensor t(shp, sv);
    const auto result = model.predict({t});
}
```

```bash
python3 fdeep_issue_209.py
python3 convert_model.py fdeep_issue_209.h5 fdeep_issue_209.json
g++ -std=c++14 -O3 fdeep_issue_209.cpp -o fdeep_issue_209
./fdeep_issue_209
```

It works without an error:

```
Loading json ... done. elapsed time: 0.000144 s
Building model ... done. elapsed time: 0.000206 s
Running test 1 of 1 ... done. elapsed time: 0.000037 s
Loading, constructing, testing of fdeep_issue_209.json took 0.000494 s overall.
```

Same with `axis=1` and `axis=2`. :man_shrugging:

Please make sure you're really using the latest version of frugally-deep from the `master` branch. Also when providing a minimal example, please check that the code actually compiles and really reproduces the error.

**Comment by mortenwp** [comment]
Part of  the mystery with the base64 values is solved: 
They appear at the bottom of the model in the trainable parameters section.
At least it makes sense to me where they are being read from

**Comment by mortenwp** [comment]
The cpp is okay on my side... the angle brackets embedded in html probably broke it

Can you try compiling with optimization. The optimizer might throw away the final lines because I dont use the result for anything

**Comment by Dobiasd** [comment]
> The cpp is okay on my side... the angle brackets embedded in html probably broke it

See here for how to post correctly formatted code in a comment on GitHub: https://help.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks

---

> Can you try compiling with optimization. The optimizer might throw away the final lines because I dont use the result for anything

OK, now without optimizations and with actually using the result, so it can not be optimized away:

```python
# fdeep_issue_209.py
import numpy as np
from tensorflow.keras.layers import Input, Permute, BatchNormalization
from tensorflow.keras.models import Model

inputs = Input(shape=(9, 16, 1))
#try different axis
x = BatchNormalization(axis=3)(inputs)
model = Model(inputs=inputs, outputs=x)

model.compile(loss='categorical_crossentropy', optimizer='nadam')
model.save('fdeep_issue_209.h5', include_optimizer=False)
```

```cpp
// fdeep_issue_209.cpp
#include <fdeep/fdeep.hpp>
int main()
{
    const auto model = fdeep::load_model("fdeep_issue_209.json");
    const auto shapes= model.get_input_shapes();
    std::vector<float> v(9*16);
    const fdeep::shared_float_vec sv(fplus::make_shared_ref<fdeep::float_vec>(v));
    fdeep::tensor_shape shp(9,16,1);
    fdeep::tensor t(shp, sv);
    const auto result = model.predict({t});
    std::cout << result[0].get(fdeep::tensor_pos(0, 0, 0)) << std::endl;
}
```

```bash
python3 fdeep_issue_209.py
python3 convert_model.py fdeep_issue_209.h5 fdeep_issue_209.json
g++ -std=c++14 fdeep_issue_209.cpp -o fdeep_issue_209
./fdeep_issue_209
```

Output:

```
Loading json ... done. elapsed time: 0.000491 s
Building model ... done. elapsed time: 0.000539 s
Running test 1 of 1 ... done. elapsed time: 0.000119 s
Loading, constructing, testing of fdeep_issue_209.json took 0.001468 s overall.
```

---

In case you can not create a valid minimal example to reproduce the problem, feel free to upload your full model, i.e., your `.h5` file, to google drive or similar, so I can check what's happening with it.

**Comment by mortenwp** [comment]
Thanks, yes without:) 
 now I cannot reproduce my earlier results, and getting problems with converting my models that I did not have before... This is probably a sign that I should stop for today. Thanks for your time, its much appreciated :)

Thanks for the link to post code snippets on github.

I will give the small test example another go tomorrow, before resorting to uploading the big model.

**Comment by mortenwp** [comment]
if i use a model that has been converted with an old version of the convert script, rather than the new version it seems that things fall overs.

The json files are slighly different, the one coming out of the old script always deals with 5 dimensions, whereas the one that comes out the new script in the small test example only deals with 3 dimension. Until now I had assumed that an old json file was just as good as a new one. 
For the small test example that we have been playing with I now get the error I see in the big example if I use the old convert model script

I will let you know how it works on my original file

**Comment by mortenwp** [comment]
If I reconvert a h5 file with the new script, and adjust my code so I dont do explicit 5 dimensional shapes but just 3 dimensional (which is more aligned with keras I guess) my code runs with no problem. 

So user error :)

If the fileformat changes it might be an idea to check for a version number, and either abort 
with a file version error or handle the different versions of the file format correctly

**Comment by Dobiasd** [comment]
Ah, cool. So maybe your model would have worked from the start if the fdeep versions to convert from h5 to json would have been the same as the one used to run in C++.

Nevertheless, the issue was not useless, because the bugfix for the permute layer way real, even though it might not have been needed for your model.

Also, your idea with making sure version match (json file and c++) sounds good to me. I'll think about a possible implementation.

Thanks. :slightly_smiling_face:
